### ISSUE TITLE
'Failed to fetch subvolume detail' on restricted mountpoint when using "backend btrfs-progs-sudo"

### ISSUE THREAD

**Issue by Omar007** [digint/btrbk#257]
First some reference on the setup:
- The program is run as a dedicated user 'btrbk'
- It is using the sudo-btrfs-progs backend for which this user is explicitly granted access.
- Subvolid=5 mountpoints located on a restricted location; only accessible for root

This worked perfectly before but with the release of v0.27.0, this setup no longer works due to the use of readlink. It will now fail with 'Failed to fetch subvolume detail' as readlink will error out when executed from the dedicated user due to the mountpoint residing on a restricted location.

**Comment by digint** [comment]
Thanks for reporting, this is a regression in v0.27.0.

I fixed this in the [fix-readlink-sudo branch](/digint/btrbk/tree/fix-readlink-sudo), in 8f475f25a09e2c6d4ecb0ca43be1de59b187b1ad: if backend is set to btrfs-progs-sudo, "readlink" will be called as "sudo -n readlink".

Could you please check if this works for you? Add "readlink" in `/etc/sudoers` and:
```
cd /tmp/
wget https://raw.githubusercontent.com/digint/btrbk/fix-readlink-sudo/btrbk
chmod +x btrbk
./btrbk dryrun -l debug
```

It would be nice to have an example setup of `/etc/sudoers` in the man pages. You seem to have a working setup, would you mind sharing `/etc/sudoers`?

**Comment by Omar007** [comment]
I'll give that branch a go this weekend.

I'd be happy to provide the sudoers file (see below). In addition to the sudoers file, the whole setup itself consists of a few more files to set up the user, home, keys etc which I may put on GitHub at a later moment.

Anyway, here is the sudoers file in advance ;)
It's not much more than a 1:1 mapping of the btrfs-progs-sudo backend cmd map.
Looking at the file I did notice I had put one command in but disabled it again. I think I put it in because it was/is used somewhere within btrbk as well but turned out to not be run using sudo.

```
# Allow users of the group btrbk to use btrfs without password
# List of commands matches btrbk's btrfs-progs-sudo backend map
%btrbk ALL=NOPASSWD:/usr/bin/btrfs subvolume list
%btrbk ALL=NOPASSWD:/usr/bin/btrfs subvolume show
%btrbk ALL=NOPASSWD:/usr/bin/btrfs subvolume snapshot
%btrbk ALL=NOPASSWD:/usr/bin/btrfs subvolume delete
#%btrbk ALL=NOPASSWD:/usr/bin/btrfs subvolume find-new
%btrbk ALL=NOPASSWD:/usr/bin/btrfs send
%btrbk ALL=NOPASSWD:/usr/bin/btrfs receive
%btrbk ALL=NOPASSWD:/usr/bin/btrfs filesystem usage
%btrbk ALL=NOPASSWD:/usr/bin/btrfs qgroup destroy
%btrbk ALL=NOPASSWD:/usr/bin/readlink -e
```
(added readlink for completeness)

**Comment by Omar007** [comment]
Had a small moment to pull up the files remotely so I threw them onto a GitHub repo just now.
https://github.com/Omar007/btrbk-contrib

The sudoers file was actually near the same as what I had posted previously, except also allowing the arguments to the commands; https://github.com/Omar007/btrbk-contrib/blob/master/btrbk.sudoers

It also listed another one in addition to 'subvolume find-new' that I apparently noticed in btrbk at some point but turned out to not be run with sudo at any point; 'filesystem df'.
I have no idea if that is a problem or not, for how I've been running btrbk it doesn't seem to have been.

**Comment by aude** [comment]
@digint, this was funny. I just experienced this problem myself, but for another reason: I tightened permissions on btrbk target "repository".

Tracked it down to `readlink` being run as an unprivileged user, and was leaning towards that it maybe should run through `sudo`. Implemented most of your commit, then came here and discovered you'd already fixed it.

Applied the changes from 8f475f2 manually, can confirm it works for me as well.
